src/pow.rs

```diff
@@ -0,0 +1,1510 @@
+    /// Creates a U256 from a big-endian array of u64's
+    fn from(a: [u64; 4]) -> Self {
+        let mut ret = U256::ZERO;
+        ret.0 = (a[0] as u128) << 64 ^ (a[1] as u128);
+        ret.1 = (a[2] as u128) << 64 ^ (a[3] as u128);
+        ret
+    }
+}
+
+/// Implements `From<$ty>` on U256 for unsigned types.
+macro_rules! impl_from_unsigned_type_for_uint256 {
+    ($ty:ident) => {
+        impl From<$ty> for U256 {
+            fn from(x: $ty) -> U256 {
+                let little = u128::from(x);
+                U256(0, little)
+            }
+        }
+    };
+}
+impl_from_unsigned_type_for_uint256!(u8);
+impl_from_unsigned_type_for_uint256!(u16);
+impl_from_unsigned_type_for_uint256!(u32);
+impl_from_unsigned_type_for_uint256!(u64);
+impl_from_unsigned_type_for_uint256!(u128);
+
+/// Implements `TryFrom<$ty>` on U256 for signed types.
+macro_rules! impl_try_from_signed_type_for_uint256 {
+    ($ty:ident) => {
+        impl TryFrom<$ty> for U256 {
+            type Error = TryFromError;
+
+            fn try_from(x: $ty) -> Result<Self, Self::Error> {
+                if x < 0 {
+                    return Err(TryFromError(i128::from(x)));
+                }
+                let little = x as u128; // Cast ok because we check sign above.
+                Ok(U256(0, little))
+            }
+        }
+    };
+}
+impl_try_from_signed_type_for_uint256!(i8);
+impl_try_from_signed_type_for_uint256!(i16);
+impl_try_from_signed_type_for_uint256!(i32);
+impl_try_from_signed_type_for_uint256!(i64);
+impl_try_from_signed_type_for_uint256!(i128);
+
+/// Error from `TryFrom<signed type>` implementations, occurs when input is negative.
+#[derive(Debug)]
+pub struct TryFromError(i128);
+
+impl fmt::Display for TryFromError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "attempt to create unsigned integer type from negative number: {}", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for TryFromError {}
+
+impl Add for U256 {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self { self.wrapping_add(rhs) }
+}
+
+impl Sub for U256 {
+    type Output = Self;
+    fn sub(self, rhs: Self) -> Self { self.wrapping_sub(rhs) }
+}
+
+impl Mul for U256 {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self { self.wrapping_mul(rhs) }
+}
+
+/// Implements mul by unsigned int in both directions. Requires `u64::from($int)`.
+macro_rules! impl_mul_for_uint256 {
+    ($int:ident) => {
+        impl Mul<$int> for U256 {
+            type Output = Self;
+            fn mul(self, rhs: $int) -> Self { self.mul_u64(u64::from(rhs)) }
+        }
+
+        impl Mul<U256> for $int {
+            type Output = U256;
+            fn mul(self, rhs: U256) -> U256 { rhs.mul_u64(u64::from(self)) }
+        }
+    };
+}
+impl_mul_for_uint256!(u8);
+impl_mul_for_uint256!(u16);
+impl_mul_for_uint256!(u32);
+impl_mul_for_uint256!(u64);
```

Comment:
Because we multiply half a 'word' at a time using stdlib integer types. mul by `u32` uses a `u64`, mul by `u64` uses a `u128` etc., to implement mul by `u128` I'd have to work out another method of multiplying - which I was too lazy to do.